core/drivers/crypto/stm32/stm32_hash.c

```diff
@@ -0,0 +1,909 @@
+		     i < ROUNDUP_DIV(c->remain.len, sizeof(uint32_t));
+		     i++) {
+			res = hash_write_data(base, c->remain.buf[i]);
+			if (res)
+				goto exit;
+			c->remain.buf[i] = 0; /* Reset to 0 */
+		}
+
+		io_clrsetbits32(base + _HASH_STR, _HASH_STR_NBLW_MASK,
+				8U * (c->remain.len % sizeof(uint32_t)));
+
+		/* No more saved data */
+		c->remain.len = 0;
+	} else {
+		io_clrbits32(base + _HASH_STR, _HASH_STR_NBLW_MASK);
+	}
+
+	io_setbits32(base + _HASH_STR, _HASH_STR_DCAL);
+
+	if (c->mode == STM32_HMAC_MODE) {
+		res = write_key(base, key, len);
+		if (res)
+			goto exit;
+	}
+
+	res = hash_get_digest(c, digest);
+
+exit:
+	clk_disable(c->dev->pdata.clock);
+	mutex_unlock(&c->dev->lock);
+
+	return res;
+}
+
+TEE_Result stm32_hash_init(struct stm32_hash_context *c, const uint8_t *key,
+			   size_t len)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	assert(c);
+
+	if ((!key || !len) && c->mode != STM32_HASH_MODE)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	mutex_lock(&c->dev->lock);
+
+	if (clk_enable(c->dev->pdata.clock)) {
+		EMSG("Fail to enable clk %s", c->dev->pdata.clock->name);
+		panic();
+	}
+
+	c->remain.len = 0;
+	/* First queue is block_size + one register */
+	c->queue_size = c->block_size + sizeof(uint32_t);
+	memset(c->remain.buf, 0, c->queue_size);
+
+	res = hw_init(c, key, len);
+	if (res)
+		goto exit;
+
+	res = save_context(c);
+
+exit:
+	clk_disable(c->dev->pdata.clock);
+	mutex_unlock(&c->dev->lock);
+
+	return res;
+}
+
+static TEE_Result stm32_hash_parse_fdt(struct stm32_hash_platdata *pdata,
+				       const void *fdt, int node,
+				       const void *compat_data)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	size_t reg_size = 0;
+	paddr_t reg = 0;
+
+	res = rstctrl_dt_get_by_index(fdt, node, 0, &pdata->reset);
+	if (res != TEE_SUCCESS && res != TEE_ERROR_ITEM_NOT_FOUND)
+		return res;
+
+	res = clk_dt_get_by_index(fdt, node, 0, &pdata->clock);
+	if (res)
+		return res;
+
+	res = fdt_reg_info(fdt, node, &reg, &reg_size);
+	if (res)
+		return res;
+
+	pdata->base = (vaddr_t)phys_to_virt(reg, MEM_AREA_IO_SEC, reg_size);
+	if (!pdata->base)
+		panic();
+
+	pdata->compat = (struct stm32_hash_compat *)compat_data;
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * Initialize HASH driver.
+ *
+ * return TEE_SUCCESS if OK.
+ */
+static TEE_Result stm32_hash_probe(const void *fdt, int node,
+				   const void *compat_data)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint32_t __maybe_unused rev = 0;
+	struct stm32_hash_platdata temp_pdata = { };
+
+	res = stm32_hash_parse_fdt(&temp_pdata, fdt, node, compat_data);
+	if (res)
+		return res;
+
+	stm32_hash = calloc(1, sizeof(*stm32_hash));
+
+	if (!stm32_hash)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	memcpy(&stm32_hash->pdata, &temp_pdata, sizeof(temp_pdata));
+
+	if (clk_enable(stm32_hash->pdata.clock)) {
+		EMSG("Fail to enable clk %s", stm32_hash->pdata.clock->name);
+		panic();
+	}
+
+	rev = io_read32(stm32_hash->pdata.base + _HASH_VERR);
+	FMSG("STM32 HASH V%u/%u", (rev & _HASH_VERR_MAJREV) >> 4,
```

Comment:
`PRIx32`
Maybe prefer "vX.Y" message over "VX/Y'.

Reply:
`rev` is of type `uint32_t`, hence `(rev & ...) >> 4` and `rev & ...` will both also be. Therefore both should use `PRIu32`.

My comment was referring to printing e.g. `(...) STM32 HASH v1.2` rather than `(...) STM32 HASH V1/2`.
That said, I have no strong opinion. At your will.

Reply:
Done, agree `v1.2` is a bit better